File: bench/cpp/records.cpp
// Variant records: see bench/goose/records_var.goose. Every row here has to
// pad each element out to the largest variant; the string payload either adds
// an allocation (std::string) or forces a fixed-capacity buffer on everyone.
//   VARIANT 0: vector<unique_ptr<Base>> with virtual dispatch (idiomatic OO).
//   VARIANT 1: vector<variant<...>> with a std::string payload.
//   VARIANT 2: vector<variant<...>> with a fixed 15-byte payload buffer.
#include "bench.h"
#include <vector>
#include <memory>
#include <string>
#include <variant>

static std::string word(uint64_t seed) {
    std::string w;
    uint64_t r = xs_next(seed);
    long long n = 4 + (long long)xs_mod(r, 12);
    for (long long i = 0; i < n; i++) { r = xs_next(r); w.push_back((char)('a' + xs_mod(r, 26))); }
    return w;
}

struct Buf {                                  // The fixed-capacity stand-in.
    unsigned char len = 0;
    char b[15];
    void set(const std::string &s) { len = (unsigned char)s.size(); for (size_t i = 0; i < s.size(); i++) b[i] = s[i]; }
};

#if VARIANT == 0
struct Base { virtual ~Base() = default; virtual long long contrib() const = 0; };
struct Tick : Base { long long contrib() const override { return 1; } };
struct Move : Base { int32_t id; int16_t dx, dy;
    long long contrib() const override { return id + dx + dy; } };
struct Say : Base { int32_t id; std::string text;
    long long contrib() const override { return id + (long long)text.size() + (unsigned char)text[0]; } };
struct Quit : Base { int32_t id; uint8_t code;
    long long contrib() const override { return id + code; } };
#else
struct Tick { };
struct Move { int32_t id; int16_t dx, dy; };
  #if VARIANT == 1
struct Say { int32_t id; std::string text; };
  #else
struct Say { int32_t id; Buf text; };
  #endif
struct Quit { int32_t id; uint8_t code; };
using Event = std::variant<Tick, Move, Say, Quit>;
#endif

int main() {
#if VARIANT == 0
    std::vector<std::unique_ptr<Base>> log;
#else
    std::vector<Event> log;
#endif
    uint64_t r = 12345;
    for (long long i = 0; i < BENCH_N; i++) {
        r = xs_next(r);
        uint64_t k = xs_mod(r, 8);
        int32_t id = (int32_t)xs_mod(r >> 3, 100000);
        if (k < 4) {
            int16_t dx = (int16_t)((long long)xs_mod(r >> 20, 201) - 100);
            int16_t dy = (int16_t)((long long)xs_mod(r >> 40, 201) - 100);
#if VARIANT == 0
            auto m = std::make_unique<Move>(); m->id = id; m->dx = dx; m->dy = dy;
            log.push_back(std::move(m));
#else
            log.push_back(Move { id, dx, dy });
#endif
        } else if (k < 6) {
            std::string t = word(r);
#if VARIANT == 0
            auto s = std::make_unique<Say>(); s->id = id; s->text = t;
            log.push_back(std::move(s));
#elif VARIANT == 1
            log.push_back(Say { id, t });
#else
            Say s; s.id = id; s.text.set(t);
            log.push_back(s);
#endif
        } else if (k < 7) {
            uint8_t code = (uint8_t)xs_mod(r >> 20, 256);
#if VARIANT == 0
            auto q = std::make_unique<Quit>(); q->id = id; q->code = code;
            log.push_back(std::move(q));
#else
            log.push_back(Quit { id, code });
#endif
        } else {
#if VARIANT == 0
            log.push_back(std::make_unique<Tick>());
#else
            log.push_back(Tick { });
#endif
        }
    }
    long long total = 0;
    for (int p = 0; p < 4; p++) {
#if VARIANT == 0
        for (const auto &e : log) total += e->contrib();
#else
        for (const auto &e : log) {
            switch (e.index()) {
                case 0: total += 1; break;
                case 1: { const Move &m = std::get<Move>(e); total += m.id + m.dx + m.dy; break; }
                case 2: { const Say &s = std::get<Say>(e);
  #if VARIANT == 1
                          total += s.id + (long long)s.text.size() + (unsigned char)s.text[0];
  #else
                          total += s.id + s.text.len + (unsigned char)s.text.b[0];
  #endif
                          break; }
                default: { const Quit &q = std::get<Quit>(e); total += q.id + q.code; break; }
            }
        }
#endif
    }
    emit((long long)log.size());
    emit(total);
    return 0;
}

File: bench/rust/records.rs
// Variant records: see bench/goose/records_var.goose. A Rust `enum` is the
// single obvious way to write this, and it is a good one -- the discriminant is
// a byte and the compiler finds a niche for it inside the `String` payload -- but
// it is still a fixed enum: every element of the `Vec` is as large as the
// largest variant, and the `Say` text is a separate heap allocation on top. That
// is the shape records_fixed.goose exists to be the control for.
mod bench;
use bench::*;

const N: i64 = 500000;                   // BENCH_N
const PASSES: i64 = 4;

enum Event {
    Tick,
    Move { id: i32, dx: i16, dy: i16 },
    Say { id: i32, text: String },
    Quit { id: i32, code: u8 },
}

// A word of 4..15 letters from its own chain, so the caller's stream advances
// once per event whatever the variant is.
fn word(seed: u64) -> String {
    let mut w = String::new();
    let mut r = xs_next(seed);
    let n = 4 + xs_mod(r, 12);
    for _ in 0..n {
        r = xs_next(r);
        w.push((b'a' + xs_mod(r, 26) as u8) as char);
    }
    w
}

fn main() {
    let mut log: Vec<Event> = Vec::new();
    let mut r: u64 = 12345;
    for _ in 0..N {
        r = xs_next(r);
        let k = xs_mod(r, 8);
        let id = xs_mod(r >> 3, 100000) as i32;
        if k < 4 {
            log.push(Event::Move { id, dx: (xs_mod(r >> 20, 201) - 100) as i16,
                                       dy: (xs_mod(r >> 40, 201) - 100) as i16 });
        } else if k < 6 {
            log.push(Event::Say { id, text: word(r) });
        } else if k < 7 {
            log.push(Event::Quit { id, code: xs_mod(r >> 20, 256) as u8 });
        } else {
            log.push(Event::Tick);
        }
    }
    let mut total: i64 = 0;
    for _ in 0..PASSES {
        for e in &log {
            total += match e {
                Event::Tick => 1,
                Event::Move { id, dx, dy } => *id as i64 + *dx as i64 + *dy as i64,
                Event::Say { id, text } => *id as i64 + text.len() as i64
                                           + text.as_bytes()[0] as i64,
                Event::Quit { id, code } => *id as i64 + *code as i64,
            };
        }
    }
    emit(log.len() as i64);
    emit(total);
}
